integration/xgboost/processor/src/nvflare-plugin/local_processor.cc

```diff
@@ -0,0 +1,326 @@
+/**
+ * Copyright 2014-2024 by XGBoost Contributors
+ */
+#include <iostream>
+#include <cstring>
+#include <cstdint>
+#include <chrono>
+#include "local_processor.h"
+#include "data_set_ids.h"
+#include "util.h"
+
+const char kParamDebug[] = "debug";
+const char kParamDamDebug[] = "dam_debug";
+const char kParamPrintTiming[] = "print_timing";
+
+void LocalProcessor::Initialize(bool active, std::map<std::string, std::string> params) {
+    active_ = active;
+    print_timing_ = get_bool(params, kParamPrintTiming);
+    debug_ = get_bool(params, kParamDebug);
+    dam_debug_ = get_bool(params, kParamDamDebug);
+}
+
+void LocalProcessor::Shutdown() {
+    gh_pairs_ = nullptr;
+    FreeEncryptedData(encrypted_gh_);
+    delete histo_;
+    cuts_.clear();
+    slots_.clear();
+}
+
+void LocalProcessor::FreeBuffer(void *buffer) {
+    free(buffer);
+}
+
+void* LocalProcessor::ProcessGHPairs(std::size_t *size, const std::vector<double>& pairs) {
+    if (debug_) {
+        std::cout << "ProcessGHPairs called with pairs size: " << pairs.size() << std::endl;
+    }
+
+    if (print_timing_) {
+        std::cout << "Encrypting " << pairs.size()/2 << " GH Pairs" << std::endl;
+    }
+    auto start = std::chrono::system_clock::now();
+
+    auto encrypted_data = EncryptVector(pairs);
+
+    if (print_timing_) {
+        auto end = std::chrono::system_clock::now();
+        auto secs = (double) std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() / 1000.0;
+        std::cout << "Encryption time: " << secs << " seconds" << std::endl;
+    }
+
+    DamEncoder encoder(kDataSetGHPairs, true, dam_debug_);
+    encoder.AddBuffer(encrypted_data);
+    auto buffer = encoder.Finish(*size);
+    FreeEncryptedData(encrypted_data);
+
+    // Save pairs for future operations. This is only called on active site
+    gh_pairs_ = new std::vector<double>(pairs);
+
+    return buffer;
```

Comment:
This buffer is dynamically allocated, and must be freed after use.

Reply:
Changed to object, not a pointer.